### PR TITLE
Handling of IOException from channel.read, channel.write

### PR DESCRIPTION
**Pull request by cmebarrow**
This is the fix for issue #9.

### CODE REVIEW

File: src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java [hunk 1]
```diff
@@ -114,7 +106,16 @@ private int handleRead() throws IOException
...
 
-            int bytesRead = channel.read(readBuffer);
+            int bytesRead;
+            try
+            {
+                bytesRead = channel.read(readBuffer);
+            }
+            catch(IOException ex)
+            {
+                // RST from client. We treat this as end of stream (see issue #9).
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java]: Please avoid referencing specific issues from the code, just document the intent instead (if not already clear from the code).
ie. `// treat TCP reset as end-of-stream`

File: src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java [hunk 2]
```diff
@@ -94,18 +94,10 @@ private Stream(
...
-                LangUtil.rethrowUnchecked(ex);
-                return 0;
-            }
+            return handleRead();
         }
 
-        private int handleRead() throws IOException
+        private int handleRead()
         {
             assert readableBytes > 0;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java]: Please remove lines `97-101`, leaving `handleStream` method in place, now that `IOException` is no longer thrown.

Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java]: done

File: src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java (hunk 1 above)
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/tcp/internal/reader/stream/StreamFactory.java]: done

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        "${streams}/client.sent.data.received.reset/client/source"
+    })
+    @BMRule(name = "processData",
+    targetClass = "^java.nio.channels.SocketChannel",
+    targetMethod = "write(java.nio.ByteBuffer)",
+    condition =
+      "callerEquals(\"org.reaktivity.nukleus.tcp.internal.writer.stream.StreamFactory$Stream.processData\", true, true)",
+      action = "throw new IOException(\"Simulating an IOException from write\")"
+    )
+    public void ioExceptionFromImmediateWriteShouldResultInReset() throws Exception
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java]: Rename test with the pattern `shouldResetWhenXXX`.

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        @BMRule(name = "handleWrite",
+        targetClass = "^java.nio.channels.SocketChannel",
+        targetMethod = "write(java.nio.ByteBuffer)",
+        condition =
+          "callerEquals(\"org.reaktivity.nukleus.tcp.internal.writer.stream.StreamFactory$Stream.handleWrite\", true, true)",
+          action = "throw new IOException(\"Simulating an IOException from write\")"
+        )
+    })
+    public void ioExceptionFromDelayedWriteShouldResultInReset() throws Exception
+    {
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java]: Rename test with the pattern `shouldResetWhenXXX`.

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    })
+    @BMRule(name = "processData",
+    targetClass = "^java.nio.channels.SocketChannel",
+    targetMethod = "write(java.nio.ByteBuffer)",
+    condition =
+      "callerEquals(\"org.reaktivity.nukleus.tcp.internal.writer.stream.StreamFactory$Stream.processData\", true, true)",
+      action = "throw new IOException(\"Simulating an IOException from write\")"
+    )
+    public void ioExceptionFromImmediateWriteShouldResultInReset() throws Exception
+    {
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java]: Rename test with the pattern `shouldResetWhenXXX`.

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        @BMRule(name = "handleWrite",
+        targetClass = "^java.nio.channels.SocketChannel",
+        targetMethod = "write(java.nio.ByteBuffer)",
+        condition =
+          "callerEquals(\"org.reaktivity.nukleus.tcp.internal.writer.stream.StreamFactory$Stream.handleWrite\", true, true)",
+          action = "throw new IOException(\"Simulating an IOException from write\")"
+        )
+    })
+    public void ioExceptionFromDelayedWriteShouldResultInReset() throws Exception
+    {
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java]: Rename test with the pattern `shouldResetWhenXXX`.

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java (hunk 3 above)
Comment (cmebarrow) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java]: done

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java (hunk 4 above)
Comment (cmebarrow) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ClientIOExceptionFromWriteIT.java]: done

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java (hunk 5 above)
Comment (cmebarrow) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java]: done

File: src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java (hunk 6 above)
Comment (cmebarrow) [inline src/test/java/org/reaktivity/nukleus/tcp/internal/streams/ServerIOExceptionFromWriteIT.java]: done

**Pull request merged**